### DISCUSSION TITLE
Tokens limits, Rebuild Index, Vector Embedding: An explanation request

### DISCUSSION THREAD

**Discussion by FamedBear16** [logancyang/obsidian-copilot#189] [Q&A]
Hi, 
I am new to this plug-in and LLM and I hope someone can clarify for me the follwoing questions.

I have created a note containing around 91000 Tokens. It is a full book converted  in plain text.
I am using the "QA: Active note" function and I have build the index using Open the AI Embedded settings in the Copilot Plug-in options

What happens when I ask a question in the "QA Acive note" mode to Chat GPT 3.5 or 4? 
Do I use all the 91000 token every time I ask one question? 
It the full note each time submitted to Open AI servers?

Can I change from Chat GPT 3.5 to 4 without rebuilding the index?

Since 90000 Tokens is above the context limit of GPT 3.5 16k does is make any difference in its usage with GPT 4 Turbo?

I hope someone can clarity what is happening under the hood.

Many thanks

**Comment by logancyang** [comment]
Hi @FamedBear16 , thanks for the questions

> What happens when I ask a question in the "QA Acive note" mode to Chat GPT 3.5 or 4?
> Do I use all the 91000 token every time I ask one question?
> It the full note each time submitted to Open AI servers?

Copilot does not send the full note to OpenAI every time. How it works is through a local RAG (Retrieval Augmented Generation), which means it splits the note into small chunks of 1000 tokens, embed each chunk and store the embeddings locally. When you ask a question, Copilot embeds your question and tries to find the relevant chunk, add the chunk to your question as context in the outgoing request to OpenAI. This way, each question is prefixed with its context chunk, the full note is never sent.

> Can I change from Chat GPT 3.5 to 4 without rebuilding the index?

Yes. The model selection dropdown in the chat window is to select the Chat Model, it has nothing to do with the index. What the Chat Model receives is your chat request containing your question + its context chunk retrieved locally from your local index. 

What's in the index is the *embeddings* of the chunks along with their original text of that chunk, and embeddings are produced by calling the Embedding Model which you can set in your Copilot settings `Embedding Provider`. You can even use a local embedding model with LocalAI (requires some dev knowledge to set it up though). For best QA quality, OpenAI as the embedding provider is still best, but it costs money to produce those embeddings.
<img width="799" alt="Screenshot 2023-11-29 at 12 45 27 PM" src="https://github.com/logancyang/obsidian-copilot/assets/4860545/fdcf3ac5-50f2-43b8-94b4-6b4e9629f5c0">

In my next updates, I will introduce new modes to enable multiple notes as context, where you can either have the full notes as context directly sent to the API in Conversation mode (since we have big context models now), or you can use RAG for super long notes to save money (downside: RAG for QA is limited by chunk retrieval relevance, so the QA quality is not going to be as good as the former approach). Still thinking about how to name these 2 modes to clearly convey what they do, since Conversation mode will have QA ability as well. But I would like to avoid names like "RAG" in the mode if possible because a lot of people don't know what it is.

**Reply by FamedBear16** [reply]
HI @logancyang,
Thanks for this detailed explanation.

 For the next update about queries multiple notes I found useful this kind of questions
1. Based on my notes in /Cards/CoreIdeas ....
2. Based on all my notes that contains the word "habits" (or the "#on/therapy")...



Regards,
